File: WordPress/src/main/java/org/wordpress/android/ui/prefs/AccountSettingsFragment.java
package org.wordpress.android.ui.prefs;

import android.annotation.SuppressLint;
import android.app.ProgressDialog;
import android.os.AsyncTask;
import android.os.Bundle;
import android.preference.Preference;
import android.preference.Preference.OnPreferenceChangeListener;
import android.preference.PreferenceFragment;
import android.text.InputType;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.coordinatorlayout.widget.CoordinatorLayout;

import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;

import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.wordpress.android.R;
import org.wordpress.android.WordPress;
import org.wordpress.android.fluxc.Dispatcher;
import org.wordpress.android.fluxc.generated.AccountActionBuilder;
import org.wordpress.android.fluxc.model.AccountModel;
import org.wordpress.android.fluxc.model.SiteModel;
import org.wordpress.android.fluxc.store.AccountStore;
import org.wordpress.android.fluxc.store.AccountStore.OnAccountChanged;
import org.wordpress.android.fluxc.store.AccountStore.PushAccountSettingsPayload;
import org.wordpress.android.fluxc.store.SiteStore;
import org.wordpress.android.ui.FullScreenDialogFragment;
import org.wordpress.android.ui.FullScreenDialogFragment.OnConfirmListener;
import org.wordpress.android.ui.accounts.signup.BaseUsernameChangerFullScreenDialogFragment;
import org.wordpress.android.ui.accounts.signup.SettingsUsernameChangerFragment;
import org.wordpress.android.util.AppLog;
import org.wordpress.android.util.AppLog.T;
import org.wordpress.android.util.NetworkUtils;
import org.wordpress.android.util.SiteUtils;
import org.wordpress.android.util.ToastUtils;
import org.wordpress.android.widgets.WPSnackbar;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import javax.inject.Inject;

@SuppressWarnings("deprecation")
public class AccountSettingsFragment extends PreferenceFragment implements OnPreferenceChangeListener,
        OnConfirmListener {
    private Preference mUsernamePreference;
    private EditTextPreferenceWithValidation mEmailPreference;
    private DetailListPreference mPrimarySitePreference;
    private EditTextPreferenceWithValidation mWebAddressPreference;
    private EditTextPreferenceWithValidation mChangePasswordPreference;
    private ProgressDialog mChangePasswordProgressDialog;
    private Snackbar mEmailSnackbar;

    @Inject Dispatcher mDispatcher;
    @Inject AccountStore mAccountStore;
    @Inject SiteStore mSiteStore;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((WordPress) getActivity().getApplication()).component().inject(this);

        setRetainInstance(true);
        addPreferencesFromResource(R.xml.account_settings);

        mUsernamePreference = findPreference(getString(R.string.pref_key_username));
        mEmailPreference = (EditTextPreferenceWithValidation) findPreference(getString(R.string.pref_key_email));
        mPrimarySitePreference = (DetailListPreference) findPreference(getString(R.string.pref_key_primary_site));
        mWebAddressPreference =
                (EditTextPreferenceWithValidation) findPreference(getString(R.string.pref_key_web_address));
        mChangePasswordPreference =
                (EditTextPreferenceWithValidation) findPreference(getString(R.string.pref_key_change_password));

        mEmailPreference.getEditText()
                        .setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_EMAIL_ADDRESS);
        mEmailPreference.setValidationType(EditTextPreferenceWithValidation.ValidationType.EMAIL);
        mWebAddressPreference.getEditText().setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_URI);
        mWebAddressPreference.setValidationType(EditTextPreferenceWithValidation.ValidationType.URL);
        mWebAddressPreference.setDialogMessage(R.string.web_address_dialog_hint);
        mChangePasswordPreference.getEditText().setInputType(InputType.TYPE_TEXT_VARIATION_PASSWORD);
        mChangePasswordPreference.setValidationType(EditTextPreferenceWithValidation.ValidationType.PASSWORD);
        mChangePasswordPreference.setDialogMessage(R.string.change_password_dialog_hint);

        mEmailPreference.setOnPreferenceChangeListener(this);
        mPrimarySitePreference.setOnPreferenceChangeListener(this);
        mWebAddressPreference.setOnPreferenceChangeListener(this);
        mChangePasswordPreference.setOnPreferenceChangeListener(this);

        setTextAlignment(mEmailPreference.getEditText());
        setTextAlignment(mWebAddressPreference.getEditText());
        setTextAlignment(mChangePasswordPreference.getEditText());

        // load site list asynchronously
        new LoadSitesTask().executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View coordinatorView = inflater.inflate(R.layout.preference_coordinator, container, false);
        CoordinatorLayout coordinator = coordinatorView.findViewById(R.id.coordinator);
        View preferenceView = super.onCreateView(inflater, coordinator, savedInstanceState);
        coordinator.addView(preferenceView);
        return coordinatorView;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        refreshAccountDetails();
    }

    @Override
    public void onResume() {
        super.onResume();
        if (NetworkUtils.isNetworkAvailable(getActivity())) {
            mDispatcher.dispatch(AccountActionBuilder.newFetchSettingsAction());
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        mDispatcher.register(this);
    }

    @Override
    public void onStop() {
        mDispatcher.unregister(this);
        super.onStop();
    }

    @Override
    public boolean onPreferenceChange(Preference preference, Object newValue) {
        if (newValue == null) {
            return false;
        }

        if (preference == mEmailPreference) {
            updateEmail(newValue.toString());
            showPendingEmailChangeSnackbar(newValue.toString());
            mEmailPreference.setEnabled(false);
            return false;
        } else if (preference == mPrimarySitePreference) {
            changePrimaryBlogPreference(Long.parseLong(newValue.toString()));
            updatePrimaryBlog(newValue.toString());
            return false;
        } else if (preference == mWebAddressPreference) {
            mWebAddressPreference.setSummary(newValue.toString());
            updateWebAddress(newValue.toString());
            return false;
        } else if (preference == mChangePasswordPreference) {
            showChangePasswordProgressDialog(true);
            updatePassword(newValue.toString());
        }

        return true;
    }

    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                getActivity().finish();
        }
        return super.onOptionsItemSelected(item);
    }

    private void setTextAlignment(EditText editText) {
        editText.setTextAlignment(View.TEXT_ALIGNMENT_VIEW_START);
    }

    private void showChangePasswordProgressDialog(boolean show) {
        if (show && mChangePasswordProgressDialog == null) {
            mChangePasswordProgressDialog = new ProgressDialog(getActivity());
            mChangePasswordProgressDialog.setCancelable(false);
            mChangePasswordProgressDialog.setIndeterminate(true);
            mChangePasswordProgressDialog.setMessage(getString(R.string.change_password_dialog_message));
            mChangePasswordProgressDialog.show();
        } else if (!show && mChangePasswordProgressDialog != null) {
            mChangePasswordProgressDialog.dismiss();
            mChangePasswordProgressDialog = null;
        }
    }

    private void refreshAccountDetails() {
        AccountModel account = mAccountStore.getAccount();
        mUsernamePreference.setSummary(account.getUserName());
        mEmailPreference.setSummary(account.getEmail());
        mWebAddressPreference.setSummary(account.getWebAddress());
        changePrimaryBlogPreference(account.getPrimarySiteId());
        checkIfEmailChangeIsPending();
        checkIfUsernameCanBeChanged();
    }

    private void checkIfEmailChangeIsPending() {
        AccountModel account = mAccountStore.getAccount();
        if (account.getPendingEmailChange()) {
            showPendingEmailChangeSnackbar(account.getNewEmail());
        } else if (mEmailSnackbar != null && mEmailSnackbar.isShown()) {
            mEmailSnackbar.dismiss();
        }
        mEmailPreference.setEnabled(!account.getPendingEmailChange());
    }

    // BaseTransientBottomBar.LENGTH_LONG is pointing to Snackabr.LENGTH_LONG which confuses checkstyle
    @SuppressLint("WrongConstant")
    private void showPendingEmailChangeSnackbar(String newEmail) {
        if (getView() != null) {
            if (mEmailSnackbar == null || !mEmailSnackbar.isShown()) {
                View.OnClickListener clickListener = new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        cancelPendingEmailChange();
                    }
                };

                mEmailSnackbar = Snackbar
                        .make(getView(), "", BaseTransientBottomBar.LENGTH_INDEFINITE)
                        .setAction(getString(R.string.button_discard), clickListener);
                TextView textView =
                        mEmailSnackbar.getView().findViewById(com.google.android.material.R.id.snackbar_text);
                textView.setMaxLines(4);
            }
            // instead of creating a new snackbar, update the current one to avoid the jumping animation
            mEmailSnackbar.setText(getString(R.string.pending_email_change_snackbar, newEmail));
            if (!mEmailSnackbar.isShown()) {
                mEmailSnackbar.show();
            }
        }
    }

    private void changePrimaryBlogPreference(long siteRemoteId) {
        mPrimarySitePreference.setValue(String.valueOf(siteRemoteId));
        SiteModel site = mSiteStore.getSiteBySiteId(siteRemoteId);
        if (site != null) {
            mPrimarySitePreference.setSummary(SiteUtils.getSiteNameOrHomeURL(site));
            mPrimarySitePreference.refreshAdapter();
        }
    }

    private void cancelPendingEmailChange() {
        PushAccountSettingsPayload payload = new PushAccountSettingsPayload();
        payload.params = new HashMap<>();
        payload.params.put("user_email_change_pending", "false");
        mDispatcher.dispatch(AccountActionBuilder.newPushSettingsAction(payload));
        if (mEmailSnackbar != null && mEmailSnackbar.isShown()) {
            mEmailSnackbar.dismiss();
        }
    }

    private void updateEmail(String newEmail) {
        PushAccountSettingsPayload payload = new PushAccountSettingsPayload();
        payload.params = new HashMap<>();
        payload.params.put("user_email", newEmail);
        mDispatcher.dispatch(AccountActionBuilder.newPushSettingsAction(payload));
    }

    private void updatePrimaryBlog(String blogId) {
        PushAccountSettingsPayload payload = new PushAccountSettingsPayload();
        payload.params = new HashMap<>();
        payload.params.put("primary_site_ID", blogId);
        mDispatcher.dispatch(AccountActionBuilder.newPushSettingsAction(payload));
    }

    public void updateWebAddress(String newWebAddress) {
        PushAccountSettingsPayload payload = new PushAccountSettingsPayload();
        payload.params = new HashMap<>();
        payload.params.put("user_URL", newWebAddress);
        mDispatcher.dispatch(AccountActionBuilder.newPushSettingsAction(payload));
    }

    public void updatePassword(String newPassword) {
        PushAccountSettingsPayload payload = new PushAccountSettingsPayload();
        payload.params = new HashMap<>();
        payload.params.put("password", newPassword);
        mDispatcher.dispatch(AccountActionBuilder.newPushSettingsAction(payload));
    }

    @SuppressWarnings("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onAccountChanged(OnAccountChanged event) {
        if (!isAdded()) {
            return;
        }

        // When account change is caused by password change, progress dialog will be shown (i.e. not null).
        if (mChangePasswordProgressDialog != null) {
            showChangePasswordProgressDialog(false);

            if (event.isError()) {
                // We usually rely on event.error.type and provide our own localized message.
                // This case is exceptional because:
                // 1. The server-side error type is generic, but patching this server-side is quite involved
                // 2. We know the error string return from the server has decent localization
                String errorMessage = !TextUtils.isEmpty(event.error.message) ? event.error.message
                        : getString(R.string.error_post_account_settings);
                ToastUtils.showToast(getActivity(), errorMessage, ToastUtils.Duration.LONG);
                AppLog.e(T.SETTINGS, event.error.message);
            } else {
                ToastUtils.showToast(getActivity(), R.string.change_password_confirmation, ToastUtils.Duration.LONG);
                refreshAccountDetails();
            }
        } else {
            if (event.isError()) {
                switch (event.error.type) {
                    case SETTINGS_FETCH_GENERIC_ERROR:
                        ToastUtils.showToast(getActivity(), R.string.error_fetch_account_settings,
                                ToastUtils.Duration.LONG);
                        break;
                    case SETTINGS_FETCH_REAUTHORIZATION_REQUIRED_ERROR:
                        ToastUtils.showToast(getActivity(), R.string.error_disabled_apis,
                                ToastUtils.Duration.LONG);
                        break;
                    case SETTINGS_POST_ERROR:
                        // We usually rely on event.error.type and provide our own localized message.
                        // This case is exceptional because:
                        // 1. The server-side error type is generic, but patching this server-side is quite involved
                        // 2. We know the error string return from the server has decent localization
                        String errorMessage = !TextUtils.isEmpty(event.error.message) ? event.error.message
                                : getString(R.string.error_post_account_settings);
                        ToastUtils.showToast(getActivity(), errorMessage, ToastUtils.Duration.LONG);
                        // we optimistically show the email change snackbar, if that request fails, we should
                        // remove the snackbar
                        checkIfEmailChangeIsPending();
                        break;
                }
            } else {
                refreshAccountDetails();
            }
        }
    }

    /**
     * If the username can be changed then the control can be clicked to open to the
     * Username Changer screen.
     */
    private void checkIfUsernameCanBeChanged() {
        AccountModel account = mAccountStore.getAccount();
        mUsernamePreference.setEnabled(account.getUsernameCanBeChanged());
        mUsernamePreference.setOnPreferenceClickListener(preference -> {
            showUsernameChangerFragment();
            return true;
        });
    }

    private void showUsernameChangerFragment() {
        AccountModel account = mAccountStore.getAccount();

        final Bundle bundle =
                SettingsUsernameChangerFragment.newBundle(account.getDisplayName(), account.getUserName());

        new FullScreenDialogFragment.Builder(getActivity())
                .setTitle(R.string.username_changer_title)
                .setAction(R.string.username_changer_action)
                .setOnConfirmListener(this)
                .setHideActivityBar(true)
                .setOnDismissListener(null)
                .setContent(SettingsUsernameChangerFragment.class, bundle)
                .build()
                .show(((AppCompatActivity) getActivity()).getSupportFragmentManager(), FullScreenDialogFragment.TAG);
    }

    @Override public void onConfirm(@Nullable Bundle result) {
        if (result != null) {
            String username = result.getString(BaseUsernameChangerFullScreenDialogFragment.RESULT_USERNAME);

            if (username != null) {
                WPSnackbar.make(getView(),
                        String.format(getString(R.string.settings_username_changer_toast_content), username),
                        Snackbar.LENGTH_LONG).show();
                mUsernamePreference.setSummary(username);
            }
        }
    }

    public static String[] getSiteNamesFromSites(List<SiteModel> sites) {
        List<String> blogNames = new ArrayList<>();
        for (SiteModel site : sites) {
            blogNames.add(SiteUtils.getSiteNameOrHomeURL(site));
        }
        return blogNames.toArray(new String[blogNames.size()]);
    }

    public static String[] getHomeURLOrHostNamesFromSites(List<SiteModel> sites) {
        List<String> urls = new ArrayList<>();
        for (SiteModel site : sites) {
            urls.add(SiteUtils.getHomeURLOrHostName(site));
        }
        return urls.toArray(new String[urls.size()]);
    }

    public static String[] getSiteIdsFromSites(List<SiteModel> sites) {
        List<String> ids = new ArrayList<>();
        for (SiteModel site : sites) {
            ids.add(String.valueOf(site.getSiteId()));
        }
        return ids.toArray(new String[ids.size()]);
    }

    /*
     * AsyncTask which loads sites from database for primary site preference
     */
    private class LoadSitesTask extends AsyncTask<Void, Void, Void> {
        @Override
        protected void onPreExecute() {
            super.onPreExecute();
        }

        @Override
        protected void onCancelled() {
            super.onCancelled();
        }

        @Override
        protected Void doInBackground(Void... params) {
            List<SiteModel> sites = mSiteStore.getSitesAccessedViaWPComRest();
            mPrimarySitePreference.setEntries(getSiteNamesFromSites(sites));
            mPrimarySitePreference.setEntryValues(getSiteIdsFromSites(sites));
            mPrimarySitePreference.setDetails(getHomeURLOrHostNamesFromSites(sites));
            return null;
        }

        @Override
        protected void onPostExecute(Void results) {
            super.onPostExecute(results);
            mPrimarySitePreference.refreshAdapter();
        }
    }
}

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsFragment.kt
@file:Suppress("DEPRECATION")

package org.wordpress.android.ui.prefs.accountsettings

import android.app.ProgressDialog
import android.os.Bundle
import android.preference.Preference
import android.preference.Preference.OnPreferenceChangeListener
import android.preference.Preference.OnPreferenceClickListener
import android.text.InputType
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ListView
import android.widget.TextView
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.view.ViewCompat
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import org.wordpress.android.R
import org.wordpress.android.WordPress
import org.wordpress.android.ui.accounts.signup.BaseUsernameChangerFullScreenDialogFragment
import org.wordpress.android.ui.pages.SnackbarMessageHolder
import org.wordpress.android.ui.prefs.DetailListPreference
import org.wordpress.android.ui.prefs.EditTextPreferenceWithValidation
import org.wordpress.android.ui.prefs.EditTextPreferenceWithValidation.ValidationType
import org.wordpress.android.ui.prefs.EditTextPreferenceWithValidation.ValidationType.EMAIL
import org.wordpress.android.ui.prefs.EditTextPreferenceWithValidation.ValidationType.PASSWORD
import org.wordpress.android.ui.prefs.EditTextPreferenceWithValidation.ValidationType.URL
import org.wordpress.android.ui.prefs.PreferenceFragmentLifeCycleOwner
import org.wordpress.android.ui.prefs.accountsettings.AccountSettingsEvent.EMAIL_CHANGED
import org.wordpress.android.ui.prefs.accountsettings.AccountSettingsEvent.PASSWORD_CHANGED
import org.wordpress.android.ui.prefs.accountsettings.AccountSettingsEvent.PRIMARY_SITE_CHANGED
import org.wordpress.android.ui.prefs.accountsettings.AccountSettingsEvent.USERNAME_CHANGED
import org.wordpress.android.ui.prefs.accountsettings.AccountSettingsEvent.USERNAME_CHANGE_SCREEN_DISMISSED
import org.wordpress.android.ui.prefs.accountsettings.AccountSettingsEvent.USERNAME_CHANGE_SCREEN_DISPLAYED
import org.wordpress.android.ui.prefs.accountsettings.AccountSettingsEvent.WEB_ADDRESS_CHANGED
import org.wordpress.android.ui.prefs.accountsettings.AccountSettingsViewModel.AccountSettingsUiState
import org.wordpress.android.ui.prefs.accountsettings.AccountSettingsViewModel.ChangePasswordSettingsUiState
import org.wordpress.android.ui.prefs.accountsettings.AccountSettingsViewModel.EmailSettingsUiState
import org.wordpress.android.ui.prefs.accountsettings.AccountSettingsViewModel.PrimarySiteSettingsUiState
import org.wordpress.android.ui.prefs.accountsettings.AccountSettingsViewModel.UserNameSettingsUiState
import org.wordpress.android.ui.utils.UiHelpers
import org.wordpress.android.util.AppLog
import org.wordpress.android.util.AppLog.T.SETTINGS
import org.wordpress.android.util.SnackbarItem
import org.wordpress.android.util.SnackbarItem.Info
import org.wordpress.android.util.SnackbarSequencer
import org.wordpress.android.util.ToastUtils
import org.wordpress.android.util.ToastUtils.Duration.LONG
import org.wordpress.android.widgets.WPSnackbar
import javax.inject.Inject

private const val SNACKBAR_NO_OF_LINES_FOUR = 4
private const val EMPTY_STRING = ""

@Suppress("DEPRECATION")
class AccountSettingsFragment : PreferenceFragmentLifeCycleOwner(),
    OnPreferenceChangeListener, OnPreferenceClickListener {
    @set:Inject
    lateinit var uiHelpers: UiHelpers

    @set:Inject
    lateinit var viewModel: AccountSettingsViewModel

    @set:Inject
    lateinit var snackbarSequencer: SnackbarSequencer

    @set:Inject
    lateinit var analyticsTracker: AccountSettingsAnalyticsTracker

    @set:Inject
    lateinit var navigationHandler: AccountSettingsNavigationHandler
    private lateinit var usernamePreference: Preference
    private lateinit var emailPreference: EditTextPreferenceWithValidation
    private lateinit var primarySitePreference: DetailListPreference
    private lateinit var webAddressPreference: EditTextPreferenceWithValidation
    private lateinit var changePasswordPreference: EditTextPreferenceWithValidation

    @Suppress("DEPRECATION")
    private var changePasswordProgressDialog: ProgressDialog? = null
    private var emailSnackbar: Snackbar? = null

    @Deprecated("Deprecated")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (activity.application as WordPress).component().inject(this)
        retainInstance = true
        addPreferencesFromResource(R.xml.account_settings)
        bindPreferences()
        setUpListeners()
        emailPreference.configure(
            inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_EMAIL_ADDRESS,
            validationType = EMAIL
        )
        webAddressPreference.configure(
            inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_URI,
            validationType = URL,
            dialogMessage = R.string.web_address_dialog_hint
        )
        changePasswordPreference.configure(
            inputType = InputType.TYPE_TEXT_VARIATION_PASSWORD,
            validationType = PASSWORD,
            dialogMessage = R.string.change_password_dialog_hint
        )
    }

    private fun bindPreferences() {
        usernamePreference = findPreference(getString(R.string.pref_key_username))
        emailPreference = findPreference(getString(R.string.pref_key_email)) as EditTextPreferenceWithValidation
        primarySitePreference = findPreference(getString(R.string.pref_key_primary_site)) as DetailListPreference
        webAddressPreference = findPreference(getString(R.string.pref_key_web_address))
                as EditTextPreferenceWithValidation
        changePasswordPreference = findPreference(getString(R.string.pref_key_change_password))
                as EditTextPreferenceWithValidation
        changePasswordPreference.summary = EMPTY_STRING
    }

    private fun setUpListeners() {
        usernamePreference.onPreferenceClickListener = this@AccountSettingsFragment
        primarySitePreference.onPreferenceChangeListener = this@AccountSettingsFragment
        emailPreference.onPreferenceChangeListener = this@AccountSettingsFragment
        webAddressPreference.onPreferenceChangeListener = this@AccountSettingsFragment
        changePasswordPreference.onPreferenceChangeListener = this@AccountSettingsFragment
    }

    private fun EditTextPreferenceWithValidation.configure(
        inputType: Int,
        validationType: ValidationType,
        dialogMessage: Int? = null
    ) {
        editText.inputType = inputType
        editText.textAlignment = View.TEXT_ALIGNMENT_VIEW_START
        setValidationType(validationType)
        dialogMessage?.let { setDialogMessage(it) }
    }

    @Deprecated("Deprecated")
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val coordinatorView = inflater.inflate(R.layout.preference_coordinator, container, false)
        val coordinator: CoordinatorLayout = coordinatorView.findViewById(R.id.coordinator)
        val preferenceView = super.onCreateView(inflater, coordinator, savedInstanceState)
        val listOfPreferences = preferenceView?.findViewById<ListView>(android.R.id.list)
        if (listOfPreferences != null) {
            ViewCompat.setNestedScrollingEnabled(listOfPreferences, true)
        }
        coordinator.addView(preferenceView)
        return coordinatorView
    }

    @Deprecated("Deprecated")
    override fun onStart() {
        super.onStart()
        observeAccountSettingsViewState()
    }

    private fun observeAccountSettingsViewState() {
        this.lifecycleScope.launchWhenStarted {
            viewModel.accountSettingsUiState.collect { updateAccountSettings(it) }
        }
    }

    private fun updateAccountSettings(accountSettingsUiState: AccountSettingsUiState) {
        updateUserNamePreferenceUi(accountSettingsUiState.userNameSettingsUiState)
        updateEmailPreferenceUi(accountSettingsUiState.emailSettingsUiState)
        webAddressPreference.summary = accountSettingsUiState.webAddressSettingsUiState.webAddress
        updatePrimarySitePreferenceUi(accountSettingsUiState.primarySiteSettingsUiState)
        updateChangePasswordPreferenceUi(accountSettingsUiState.changePasswordSettingsUiState)
        accountSettingsUiState.toastMessage?.let {
            showToastMessage(it)
        }
    }

    private fun updateChangePasswordPreferenceUi(changePasswordSettingsUiState: ChangePasswordSettingsUiState) {
        showChangePasswordProgressDialog(changePasswordSettingsUiState.showChangePasswordProgressDialog)
    }

    private fun updateUserNamePreferenceUi(userNameSettingUiState: UserNameSettingsUiState) {
        usernamePreference.apply {
            summary = userNameSettingUiState.userName
            isEnabled = userNameSettingUiState.canUserNameBeChanged
        }
        if (userNameSettingUiState.showUserNameConfirmedSnackBar) {
            showUserNameSnackBar(userNameSettingUiState.newUserChangeConfirmedSnackBarMessageHolder)
            viewModel.onUserConfirmedSnackBarShown()
        }
    }

    private fun updateEmailPreferenceUi(emailSettingsUiState: EmailSettingsUiState) {
        emailPreference.apply {
            summary = emailSettingsUiState.email
            isEnabled = emailSettingsUiState.hasPendingEmailChange.not()
        }
        if (emailSettingsUiState.hasPendingEmailChange) {
            showSnackBar(emailSettingsUiState.emailVerificationMsgSnackBarMessageHolder)
        } else {
            dismissSnackBar()
        }
    }

    private fun updatePrimarySitePreferenceUi(primarySiteSettingsUiState: PrimarySiteSettingsUiState?) {
        primarySiteSettingsUiState?.let { state ->
            primarySitePreference.apply {
                value = (state.primarySite?.siteId ?: "").toString()
                summary = state.primarySite?.siteName
                entries = state.siteNames
                entryValues = state.siteIds
                canShowDialog = state.canShowChoosePrimarySiteDialog
                setDetails(state.homeURLOrHostNames)
                refreshAdapter()
                 // Add click listener to show toast
                setOnPreferenceClickListener {
                    if (state.sites?.size == 1) {
                        val message = getString(R.string.only_one_primary_site_message)
                        showToastMessage(message)
                    }
                    true
                }
            }
        } ?: run {
            primarySitePreference.apply {
                refreshAdapter()
            }
        }
    }

    private fun showToastMessage(toastMessage: String) {
        ToastUtils.showToast(activity, toastMessage, LONG)
        viewModel.onToastShown(toastMessage)
        AppLog.e(SETTINGS, toastMessage)
    }

    private fun showUserNameSnackBar(userName: SnackbarMessageHolder) {
        view?.let {
            snackbarSequencer.enqueue(
                SnackbarItem(Info(view = it, textRes = userName.message, duration = userName.duration))
            )
        }
    }

    private fun showSnackBar(snackBarMessage: SnackbarMessageHolder) {
        emailSnackbar?.let {
            if (!it.isShown) {
                emailSnackbar?.show()
            }
        } ?: run {
            view?.let {
                emailSnackbar = WPSnackbar.make(
                    it,
                    uiHelpers.getTextOfUiString(context, snackBarMessage.message),
                    BaseTransientBottomBar.LENGTH_INDEFINITE
                )
                snackBarMessage.buttonTitle?.let {
                    emailSnackbar?.setAction(
                        uiHelpers.getTextOfUiString(context, snackBarMessage.buttonTitle)
                    ) { snackBarMessage.buttonAction }
                }
                val textView = emailSnackbar?.view
                    ?.findViewById<TextView>(com.google.android.material.R.id.snackbar_text)
                textView?.maxLines = SNACKBAR_NO_OF_LINES_FOUR
            }
        }
    }

    private fun dismissSnackBar() {
        emailSnackbar?.dismiss()
    }

    @Deprecated("Deprecated")
    override fun onPreferenceClick(preference: Preference?): Boolean {
        if (preference == usernamePreference) {
            val onUserNameConfirmed = { result: Bundle? ->
                result?.getString(BaseUsernameChangerFullScreenDialogFragment.RESULT_USERNAME)?.let {
                    viewModel.onUsernameChangeConfirmedFromServer(it)
                    analyticsTracker.track(USERNAME_CHANGED)
                }
            }
            navigationHandler.showUsernameChangerScreen(
                activity,
                UserNameChangeScreenRequest(
                    userName = viewModel.accountSettingsUiState.value.userNameSettingsUiState.userName,
                    displayName = viewModel.accountSettingsUiState.value.userNameSettingsUiState.displayName,
                    onConfirm = onUserNameConfirmed,
                    onShown = { analyticsTracker.track(USERNAME_CHANGE_SCREEN_DISPLAYED) },
                    onDismiss = { analyticsTracker.track(USERNAME_CHANGE_SCREEN_DISMISSED) }
                )
            )
        }
        return true
    }

    @Deprecated("Deprecated")
    override fun onPreferenceChange(preference: Preference, newValue: Any): Boolean {
        var action: AccountSettingsEvent? = null
        when (preference) {
            emailPreference -> {
                viewModel.onEmailChanged(newValue.toString())
                action = EMAIL_CHANGED
            }
            primarySitePreference -> {
                viewModel.onPrimarySiteChanged(newValue.toString().toLong())
                action = PRIMARY_SITE_CHANGED
            }
            webAddressPreference -> {
                viewModel.onWebAddressChanged(newValue.toString())
                action = WEB_ADDRESS_CHANGED
            }
            changePasswordPreference -> {
                viewModel.onPasswordChanged(newValue.toString())
                action = PASSWORD_CHANGED
            }
        }
        if (!preference.summary.toString().equals(newValue.toString(), ignoreCase = true)) {
            action?.let { analyticsTracker.track(it) }
        }
        return true
    }

    @Deprecated("Deprecated")
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> activity.finish()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun showChangePasswordProgressDialog(show: Boolean) {
        if (!show) {
            changePasswordProgressDialog?.dismiss()
            changePasswordProgressDialog = null
            return
        }
        createChangePasswordDialogIfNull()
        changePasswordProgressDialog?.show()
    }

    @Suppress("DEPRECATION")
    private fun createChangePasswordDialogIfNull() {
        if (changePasswordProgressDialog == null) {
            changePasswordProgressDialog = ProgressDialog(activity).apply {
                setCancelable(false)
                isIndeterminate = true
                setMessage(getString(R.string.change_password_dialog_message))
            }
        }
    }
}
